File: scolioscan-expo/modules/on-device-pose/android/src/main/java/expo/modules/ondevicepose/OnDevicePoseModule.kt
package expo.modules.ondevicepose

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.net.Uri
import androidx.exifinterface.media.ExifInterface
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import com.google.mlkit.vision.pose.Pose
import com.google.mlkit.vision.pose.PoseDetection
import com.google.mlkit.vision.pose.PoseLandmark
import com.google.mlkit.vision.pose.defaults.PoseDetectorOptions
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.io.FileOutputStream

class OnDevicePoseModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("OnDevicePose")

    AsyncFunction("detectPoseOnDevice") { imageUri: String, promise: Promise ->
      val context = appContext.reactContext ?: throw Exceptions.ReactContextLost()
      // 카메라 JPEG의 EXIF 방향을 실제 Bitmap에 반영해 Pose 좌표와 JS crop 좌표계를 맞춘다.
      val image = InputImage.fromBitmap(loadExifNormalizedBitmap(context, Uri.parse(imageUri)), 0)
      val options = PoseDetectorOptions.Builder()
        .setDetectorMode(PoseDetectorOptions.SINGLE_IMAGE_MODE)
        .build()
      val poseDetector = PoseDetection.getClient(options)
      val faceOptions = FaceDetectorOptions.Builder()
        .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
        .setMinFaceSize(0.1f)
        .build()
      val faceDetector = FaceDetection.getClient(faceOptions)

      poseDetector.process(image)
        .addOnSuccessListener { pose ->
          faceDetector.process(image)
            .addOnSuccessListener { faces ->
              promise.resolve(buildResponse(pose, image.width, image.height, faces))
              poseDetector.close()
              faceDetector.close()
            }
            .addOnFailureListener { error ->
              poseDetector.close()
              faceDetector.close()
              promise.reject("ERR_ON_DEVICE_FACE", error.message ?: "온디바이스 얼굴 분석에 실패했습니다.", error)
            }
        }
        .addOnFailureListener { error ->
          poseDetector.close()
          faceDetector.close()
          promise.reject("ERR_ON_DEVICE_POSE", error.message ?: "온디바이스 포즈 분석에 실패했습니다.", error)
        }
    }

    AsyncFunction("normalizeImageForPose") { imageUri: String, promise: Promise ->
      try {
        val context = appContext.reactContext ?: throw Exceptions.ReactContextLost()
        val bitmap = loadExifNormalizedBitmap(context, Uri.parse(imageUri))
        val outputFile = File.createTempFile("pose-normalized-", ".jpg", context.cacheDir)

        FileOutputStream(outputFile).use { output ->
          if (!bitmap.compress(Bitmap.CompressFormat.JPEG, 100, output)) {
            throw IllegalStateException("Unable to save normalized image bitmap")
          }
        }

        promise.resolve(
          mapOf(
            "uri" to Uri.fromFile(outputFile).toString(),
            "width" to bitmap.width,
            "height" to bitmap.height,
          )
        )
      } catch (error: Exception) {
        promise.reject("ERR_NORMALIZE_IMAGE", error.message ?: "Failed to normalize image", error)
      }
    }
  }

  private fun loadExifNormalizedBitmap(context: Context, uri: Uri): Bitmap {
    val orientation = context.contentResolver.openInputStream(uri)?.use { input ->
      ExifInterface(input).getAttributeInt(
        ExifInterface.TAG_ORIENTATION,
        ExifInterface.ORIENTATION_NORMAL,
      )
    } ?: ExifInterface.ORIENTATION_NORMAL

    val bitmap = context.contentResolver.openInputStream(uri)?.use { input ->
      BitmapFactory.decodeStream(input)
    } ?: throw IllegalArgumentException("Unable to decode image bitmap")

    return applyExifOrientation(bitmap, orientation)
  }

  private fun applyExifOrientation(bitmap: Bitmap, orientation: Int): Bitmap {
    val matrix = Matrix()

    when (orientation) {
      ExifInterface.ORIENTATION_FLIP_HORIZONTAL -> matrix.setScale(-1f, 1f)
      ExifInterface.ORIENTATION_ROTATE_180 -> matrix.setRotate(180f)
      ExifInterface.ORIENTATION_FLIP_VERTICAL -> matrix.setScale(1f, -1f)
      ExifInterface.ORIENTATION_TRANSPOSE -> {
        matrix.setRotate(90f)
        matrix.postScale(-1f, 1f)
      }
      ExifInterface.ORIENTATION_ROTATE_90 -> matrix.setRotate(90f)
      ExifInterface.ORIENTATION_TRANSVERSE -> {
        matrix.setRotate(-90f)
        matrix.postScale(-1f, 1f)
      }
      ExifInterface.ORIENTATION_ROTATE_270 -> matrix.setRotate(-90f)
      else -> return bitmap
    }

    return Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
  }

  private fun buildResponse(pose: Pose, imageWidth: Int, imageHeight: Int, faces: List<Face>): Map<String, Any?> {
    val landmarks = landmarkTypes.map { type ->
      val landmark = pose.getPoseLandmark(type)
      toLandmarkMap(landmark, imageWidth, imageHeight)
    }
    val detected = landmarks.any { landmark ->
      (landmark["visibility"] as Double) >= PERSON_DETECTED_VISIBILITY
    }
    val faceCount = faces.size
    val faceDetected = faceCount > 0
    val faceScore = if (faceDetected) 1.0 else 0.0

    return mapOf(
      "detected" to detected,
      "landmarks" to if (detected) landmarks else null,
      "face_detected" to faceDetected,
      "face_score" to faceScore,
      "face_count" to faceCount,
    )
  }

  private fun toLandmarkMap(landmark: PoseLandmark?, imageWidth: Int, imageHeight: Int): Map<String, Double> {
    if (landmark == null || imageWidth <= 0 || imageHeight <= 0) {
      return emptyLandmark()
    }

    val position = landmark.position
    val position3D = landmark.position3D

    // 기존 서버 응답과 맞추기 위해 이미지 좌표를 0..1 정규화 좌표로 변환한다.
    return mapOf(
      "x" to clamp01(position.x.toDouble() / imageWidth.toDouble()),
      "y" to clamp01(position.y.toDouble() / imageHeight.toDouble()),
      "z" to position3D.z.toDouble(),
      "visibility" to landmark.inFrameLikelihood.toDouble(),
    )
  }

  private fun emptyLandmark(): Map<String, Double> {
    return mapOf(
      "x" to 0.0,
      "y" to 0.0,
      "z" to 0.0,
      "visibility" to 0.0,
    )
  }

  private fun clamp01(value: Double): Double {
    return value.coerceIn(0.0, 1.0)
  }

  companion object {
    private const val PERSON_DETECTED_VISIBILITY = 0.2

    private val landmarkTypes = listOf(
      PoseLandmark.NOSE,
      PoseLandmark.LEFT_EYE_INNER,
      PoseLandmark.LEFT_EYE,
      PoseLandmark.LEFT_EYE_OUTER,
      PoseLandmark.RIGHT_EYE_INNER,
      PoseLandmark.RIGHT_EYE,
      PoseLandmark.RIGHT_EYE_OUTER,
      PoseLandmark.LEFT_EAR,
      PoseLandmark.RIGHT_EAR,
      PoseLandmark.LEFT_MOUTH,
      PoseLandmark.RIGHT_MOUTH,
      PoseLandmark.LEFT_SHOULDER,
      PoseLandmark.RIGHT_SHOULDER,
      PoseLandmark.LEFT_ELBOW,
      PoseLandmark.RIGHT_ELBOW,
      PoseLandmark.LEFT_WRIST,
      PoseLandmark.RIGHT_WRIST,
      PoseLandmark.LEFT_PINKY,
      PoseLandmark.RIGHT_PINKY,
      PoseLandmark.LEFT_INDEX,
      PoseLandmark.RIGHT_INDEX,
      PoseLandmark.LEFT_THUMB,
      PoseLandmark.RIGHT_THUMB,
      PoseLandmark.LEFT_HIP,
      PoseLandmark.RIGHT_HIP,
      PoseLandmark.LEFT_KNEE,
      PoseLandmark.RIGHT_KNEE,
      PoseLandmark.LEFT_ANKLE,
      PoseLandmark.RIGHT_ANKLE,
      PoseLandmark.LEFT_HEEL,
      PoseLandmark.RIGHT_HEEL,
      PoseLandmark.LEFT_FOOT_INDEX,
      PoseLandmark.RIGHT_FOOT_INDEX,
    )
  }
}

File: scolioscan-expo/modules/on-device-pose/ios/OnDevicePoseModule.swift
import ExpoModulesCore
import MLKitPoseDetection
import MLKitFaceDetection
import MLKitVision
import UIKit

public class OnDevicePoseModule: Module {
  public func definition() -> ModuleDefinition {
    Name("OnDevicePose")

    AsyncFunction("detectPoseOnDevice") { (imageUri: String) async throws -> [String: Any] in
      guard let url = URL(string: imageUri),
            let data = try? Data(contentsOf: url),
            let image = UIImage(data: data) else {
        throw NSError(
          domain: "OnDevicePose",
          code: 1,
          userInfo: [NSLocalizedDescriptionKey: "Failed to load image from URI"]
        )
      }

      let visionImage = VisionImage(image: image)
      visionImage.orientation = image.imageOrientation

      let poseOptions = PoseDetectorOptions()
      poseOptions.detectorMode = .singleImage
      let poseDetector = PoseDetector.poseDetector(options: poseOptions)

      let faceOptions = FaceDetectorOptions()
      faceOptions.performanceMode = .fast
      faceOptions.minFaceSize = 0.1
      let faceDetector = FaceDetector.faceDetector(options: faceOptions)

      async let poseResult = self.processPose(detector: poseDetector, image: visionImage)
      async let faceResult = self.processFaces(detector: faceDetector, image: visionImage)

      let poses = try await poseResult
      let faces = try await faceResult
      let isNormalizedPoseImage = url.lastPathComponent.hasPrefix("pose-normalized-") && image.imageOrientation == .up

      return self.buildResponse(
        poses: poses ?? [],
        imageWidth: image.size.width,
        imageHeight: image.size.height,
        isNormalizedPoseImage: isNormalizedPoseImage,
        faces: faces ?? []
      )
    }

    AsyncFunction("normalizeImageForPose") { (imageUri: String) async throws -> [String: Any] in
      guard let url = URL(string: imageUri),
            let data = try? Data(contentsOf: url),
            let image = UIImage(data: data),
            let cgImage = image.cgImage else {
        throw NSError(
          domain: "OnDevicePose",
          code: 2,
          userInfo: [NSLocalizedDescriptionKey: "Failed to load image for normalization"]
        )
      }

      let swapsWidthAndHeight: Bool
      switch image.imageOrientation {
      case .left, .leftMirrored, .right, .rightMirrored:
        swapsWidthAndHeight = true
      default:
        swapsWidthAndHeight = false
      }

      let normalizedSize = CGSize(
        width: CGFloat(swapsWidthAndHeight ? cgImage.height : cgImage.width),
        height: CGFloat(swapsWidthAndHeight ? cgImage.width : cgImage.height)
      )

      // 기기 화면 배율과 무관하게 원본 픽셀 크기로 렌더링하고 EXIF 방향을 실제 픽셀에 반영한다.
      let format = UIGraphicsImageRendererFormat()
      format.scale = 1
      format.opaque = true
      let renderer = UIGraphicsImageRenderer(size: normalizedSize, format: format)
      let normalizedData = renderer.jpegData(withCompressionQuality: 1) { _ in
        image.draw(in: CGRect(origin: .zero, size: normalizedSize))
      }

      let outputUrl = FileManager.default.temporaryDirectory
        .appendingPathComponent("pose-normalized-\(UUID().uuidString).jpg")
      try normalizedData.write(to: outputUrl, options: .atomic)

      return [
        "uri": outputUrl.absoluteString,
        "width": Int(normalizedSize.width),
        "height": Int(normalizedSize.height)
      ]
    }
  }

  private func processPose(detector: PoseDetector, image: VisionImage) async throws -> [Pose]? {
  return try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<[Pose]?, Error>) in
    detector.process(image) { results, error in
      if let error = error {
        continuation.resume(throwing: error)
      } else {
        continuation.resume(returning: results)
      }
    }
  }
}

  private func processFaces(detector: FaceDetector, image: VisionImage) async throws -> [Face]? {
    return try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<[Face]?, Error>) in
      detector.process(image) { results, error in
        if let error = error {
          continuation.resume(throwing: error)
        } else {
          continuation.resume(returning: results)
        }
      }
    }
  }

  private func buildResponse(
    poses: [Pose],
    imageWidth: CGFloat,
    imageHeight: CGFloat,
    isNormalizedPoseImage: Bool,
    faces: [Face]
  ) -> [String: Any] {
    let faceCount = faces.count
    let faceDetected = faceCount > 0
    let faceScore = faceDetected ? 1.0 : 0.0

    guard let pose = poses.first, imageWidth > 0, imageHeight > 0 else {
      return [
        "detected": false,
        "landmarks": NSNull(),
        "face_detected": faceDetected,
        "face_score": faceScore,
        "face_count": faceCount
      ]
    }

    let landmarkTypes: [(name: String, type: PoseLandmarkType)] = [
      ("nose", .nose),

      ("leftEyeInner", .leftEyeInner),
      ("leftEye", .leftEye),
      ("leftEyeOuter", .leftEyeOuter),

      ("rightEyeInner", .rightEyeInner),
      ("rightEye", .rightEye),
      ("rightEyeOuter", .rightEyeOuter),

      ("leftEar", .leftEar),
      ("rightEar", .rightEar),

      ("leftMouth", .mouthLeft),
      ("rightMouth", .mouthRight),

      ("leftShoulder", .leftShoulder),
      ("rightShoulder", .rightShoulder),

      ("leftElbow", .leftElbow),
      ("rightElbow", .rightElbow),

      ("leftWrist", .leftWrist),
      ("rightWrist", .rightWrist),

      ("leftPinky", .leftPinkyFinger),
      ("rightPinky", .rightPinkyFinger),

      ("leftIndex", .leftIndexFinger),
      ("rightIndex", .rightIndexFinger),

      ("leftThumb", .leftThumb),
      ("rightThumb", .rightThumb),

      ("leftHip", .leftHip),
      ("rightHip", .rightHip),

      ("leftKnee", .leftKnee),
      ("rightKnee", .rightKnee),

      ("leftAnkle", .leftAnkle),
      ("rightAnkle", .rightAnkle),

      ("leftHeel", .leftHeel),
      ("rightHeel", .rightHeel),

      ("leftFootIndex", .leftToe),
      ("rightFootIndex", .rightToe)
    ]

    let landmarks: [[String: Any]] = landmarkTypes.map { item in
      let landmark = pose.landmark(ofType: item.type)
      let position = landmark.position
      let normalizedX: CGFloat
      let normalizedY: CGFloat

      if isNormalizedPoseImage {
        // crop용 정규화 이미지는 이미 세로 방향이므로 좌표를 다시 회전하지 않는다.
        normalizedX = position.x / imageWidth
        normalizedY = position.y / imageHeight
      } else {
        // 기존 iOS 촬영 이미지의 EXIF 방향 보정은 유지해 가이드 판정 동작을 바꾸지 않는다.
        normalizedX = 1 - (position.y / imageWidth)
        normalizedY = position.x / imageHeight
      }
 
      return [
        "name": item.name,
        "x": Double(max(0, min(1, normalizedX))),
        "y": Double(max(0, min(1, normalizedY))),
        "z": Double(position.z),
        "visibility": Double(landmark.inFrameLikelihood)
      ]
    }

    let personDetectedVisibility = 0.2
    let detected = landmarks.contains { landmark in
      (landmark["visibility"] as? Double ?? 0.0) >= personDetectedVisibility
    }

    return [
      "detected": detected,
      "landmarks": detected ? landmarks : NSNull(),
      "face_detected": faceDetected,
      "face_score": faceScore,
      "face_count": faceCount
    ]
  }
}
